lib/httplib/connstate.go

```diff
@@ -0,0 +1,237 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package httplib
+
+import (
+	"net"
+	"net/http"
+	"sync"
+)
+
+// AllConnStates indicates a subscription to all http.ConnState events
+// in ConnStateEvents and ServerConnState. Pass to the Notify
+// function of each to subscribe to all events.
+const AllConnStates = -1
+
+// TerminalConnState returns true when state is terminal.
+func TerminalConnState(state http.ConnState) bool {
+	return state == http.StateHijacked || state == http.StateClosed
```

Comment:
Should `StateHijacked` be a terminal state though? IIRC you can hijack a connection on the ResponseWriter in which case the caller basically takes over the connection completely. It is used during websocket upgrade for example. But the connection is still open so we shouldn't consider it done until it's actually closed?

I would test what happens with an application that supports websockets ([nodered](https://nodered.org/docs/getting-started/docker) is one that is very easy to run).

Reply:
Go documents consider Hijacked terminal: `This is a terminal state. It does not transition to StateClosed.`. Once it hands the connection over it no longer tracks it and no state events will fire. This is why I'm passing back the terminal state in the channel for all subscriptions, so the caller knows which terminal state occurred.

Reply:
Hmm interesting. I would definitely explain it with a comment here and also let's please make sure websockets don't break like I mentioned above.